### PR TITLE
fix issue #974 and some other warnings

### PR DESCRIPTION
**Pull request by MohammadAlTurany**
fix issue #974 and few other warnings

### CODE REVIEW

File: base/sim/FairGenericStack.h [hunk 1]
```diff
@@ -98,8 +98,11 @@ class FairGenericStack : public TVirtualMCStack
     virtual void Register() {}
 
     virtual TClonesArray* GetListOfParticles() { return 0; }
-    virtual void SetParticleArray(TClonesArray* partArray) {}
-    virtual void SetParticleArray(TClonesArray* partArray, Int_t partFrom, Int_t partTo) {}
+    virtual void SetParticleArray(__attribute__((unused)) TClonesArray* partArray) {}
+    virtual void SetParticleArray(__attribute__((unused)) TClonesArray* partArray,
+                                  __attribute__((unused)) Int_t partFrom,
+                                  __attribute__((unused)) Int_t partTo)
+    {}
```
Comment (mattkretz) [inline base/sim/FairGenericStack.h:105]: use `[[gnu::unused]]` or `[[maybe_unused]]` instead of non-standard attribute syntax

Comment (fuhlig1) [reply base/sim/FairGenericStack.h:105]: or use

 virtual void SetParticleArray(TClonesArray*) {}
virtual void SetParticleArray(TClonesArray*, Int_t, Int_t) {}

Comment (MohammadAlTurany) [reply base/sim/FairGenericStack.h:105]: Thanks! Ok I will use **[[maybe_unused]]**

Comment (mattkretz) [reply base/sim/FairGenericStack.h:105]: Note that `[[maybe_unused]]` is C++17

File: examples/simulation/Tutorial4/src/tools/FairTutorialDet4GeoHandler.cxx [hunk 2]
```diff
@@ -102,7 +102,7 @@ Int_t FairTutorialDet4GeoHandler::VolId(const Text_t* name) const
         if (name[len] != ' ') {
             return VolIdGeo(name);
         }
-        strncpy(sname, name, len);
+        memcpy(sname, name, len);
```
Comment (dennisklein) [inline examples/simulation/Tutorial4/src/tools/FairTutorialDet4GeoHandler.cxx:105]: There are actually two more issues present in this code:
* buffer overflow if `name` is longer than 20 characters
* buffer underflow if `name` is empty

I suggest something like this (untested):

```diff
@@ -100,8 +100,5 @@
-        char sname[20];
-        Int_t len = strlen(name) - 1;
-        if (name[len] != ' ') {
-            return VolIdGeo(name);
-        }
-        strncpy(sname, name, len);
-        memcpy(sname, name, len);
-        sname[len] = 0;
-        return VolIdGeo(sname);
+        std::string volname(name);
+        while (!volname.empty() && volname.back() == ' ') {
+          volname.pop_back();
+        }
+        return VolIdGeo(volname.c_str());
```

This also does a proper `rtrim` which I assume was the original intention anyways (if not, replace the `while` with an `if`).

File: base/sim/FairGenericStack.h (hunk 1 above)
Comment (MohammadAlTurany) [reply base/sim/FairGenericStack.h:105]: > or use
> 
> virtual void SetParticleArray(TClonesArray*) {}
> virtual void SetParticleArray(TClonesArray*, Int_t, Int_t) {}

In some cases it did not work!

File: examples/simulation/Tutorial4/src/tools/FairTutorialDet4GeoHandler.cxx (hunk 2 above)
Comment (MohammadAlTurany) [reply examples/simulation/Tutorial4/src/tools/FairTutorialDet4GeoHandler.cxx]: Done

File: base/sim/FairGenericStack.h (hunk 1 above)
Comment (MohammadAlTurany) [reply base/sim/FairGenericStack.h:105]: I change it to  gnu-unused as we are still requiring C++11 and not 17.

File: base/sim/fastsim/FairFastSimModel.cxx [hunk 3]
```diff
@@ -97,7 +97,7 @@ void FairFastSimModel::DoIt(const G4FastTrack& fastTrack, G4FastStep& fastStep)
             particle->GetPolarisation(polVect);
             G4ThreeVector pol(polVect.X(), polVect.Y(), polVect.Z());   // get polarisation
             G4double tim = particle->T() * 10.e9;                       // change from ns to s
-            G4double len = 0.;
+                                                                        //   G4double len = 0.;
```
Comment (mattkretz) [inline base/sim/fastsim/FairFastSimModel.cxx:100]: If `len` is unused, better remove it. The next reader of the code will be rather puzzled, trying to understand the comment. :wink:

Comment (MohammadAlTurany) [reply base/sim/fastsim/FairFastSimModel.cxx:100]: removed in #979

**Pull request merged**